lib/vanagon/component.rb

```diff
@@ -290,6 +290,10 @@ def get_build_dir
       end
     end
 
+    def get_dependency_hash
+      { name => { 'version' => version, 'ref' => options[:ref] }.delete_if { |k, v| !v } }
```

Comment:
I understand the desire to generalize this, but this entire method is self-contained and only operating on 2 elements.

If this method was taking input from outside and needed to be agnostic to the actual values its operating on, something that's general-purpose like this would be useful.

Arguments about performance, and how many hashes it may make, and how many conditionals it uses are valid, but at the end of the day there's no way around some of these things.

We still need conditionals to only insert elements if they're non-nil.  The simplest conditional is an `if` statement, so start with that.

Regarding how many hashes are created in the process, do we know the implementation of `delete_if`?  Maybe it creates an internal hash to modify so it doesn't get concurrent modification problems when attempting to iterate and remove elements at the same time.

Using `!` to test for nil isn't exactly obvious either, since that's usually something used for boolean values.  Why not use the `.nil?` method in this case?

Just because we can write it tersely doesn't mean the performance is better or that it's easier to understand what's going on.

> I wouldn't even purge empty values since JSON serialization understands nil as null

I wouldn't necessarily assume that this method will be only used for generating JSON.  We almost had it generate YAML, and it may be used by other methods that may do something different with the hash.

I would keep all concerns specific to JSON in the one method that actually writes it out as JSON.

> Also, I strongly disagree that the suggested oneliner is unreadable. That's a very, very common Ruby construction.

I strongly disagree that this is the most understandable way to write this.  Being a common Ruby construct doesn't make it more understandable.

And after all that, while there are many ways we can achieve the desired hash in Ruby, we need to move forward with the rest of the PEGI effort and not be tempted to rewrite this over and over.

I'd say we've all probably learned something from these discussions, and we're all the better for it, but time is of the essence!

We're nearly done with this change, so let's stay focused and finish whatever else is necessary to get this across the line.